geoips/interfaces/base.py

```diff
@@ -292,12 +307,99 @@ def __repr__(self):
 
 
 class BaseModulePlugin:
-    """Base class for GeoIPS plugins."""
+    """Base class for Module-Based GeoIPS plugins."""
 
     pass
 
 
-class BaseInterface:
+class BaseTextPlugin(ABC):
+    """Base class for Text-Based GeoIPS plugins."""
+
+    def __init__(self, interface, plugin_name, **kwargs):
+        """Abstract intitialization method used for the inheriting text-plugin class.
+
+        Required to initialize the inheriting text plugin into an actual object..
+
+        Parameters
+        ----------
+        interface: Child of BaseTextInteface
+            - The corresponding interface to retrieve the plugin from
+        plugin_name: str
+            - The name of the plugin to initialize
+        kwargs: dict
+            - A dictionary of optional keyword arguments relative to the later-to-be
+              defined __init__ function.
+        """
+        self._validate(interface, plugin_name)
+
+    @abstractmethod
+    def __call__(self, **kwargs):
+        """Abstract call method used for the inheriting text-plugin class.
+
+        This is what will be called when a text plugin of some type is actually invoked.
+
+        Parameters
+        ----------
+        kwargs: dict
+            - A dictionary of optional keyword arguments relative to the later-to-be
+              defined __call__ function.
+        """
+        pass
+
+    def __repr__(self):
+        """Unambiguous representation of the Text Plugin for dev purposes."""
+        return f"{self.__class__.__name__}('{self.name}')"
+
+    def __str__(self):
+        """Readable representation of the Text Plugin."""
+        return f"{self.__class__.__name__}: '{self.name}'"
+
+    def _validate(self, interface, plugin_name):
+        """Validate the plugin from the provided interface found under 'plugin_name'.
+
+        Make sure that plugin found under 'plugin_name' has attributes:
+            - ["interface", "family", "package", "relpath", "docstring", "plugin_type"]
+
+        Parameters
+        ----------
+        interface: Child of BaseTextInteface
+            - The corresponding interface to retrieve the plugin from
+        plugin_name: str
+            - The name of the plugin to validate
+
+        Possibly Raises
+        ---------------
+        error: PluginRegistryError
+            - Will be raised if the plugin is missing a required attribute
+
+        Possibly Throws
```

Comment:
# Raise vs throw in Python
In python, we can only `raise` errors; there is no `throw`. 

Well, I guess there's the [`throw()` in generators](https://python-reference.readthedocs.io/en/latest/docs/generator/throw.html) but the docs point out `throw()` is really a wrapper for `raise`. So it's `raise` all the way down. Except when there's a namespace conflict lol.

This is a bit confusing, because we `catch` exceptions so .... don't we throw them? Well, we don't _really_ `catch` exceptions, we `except` them. Eg.

``` python
try: 
    climb_mount_everest()
except CantBreath:
	print("Sorry, you died :)")
	exit(1)
```

Other programming languages (eg. Ruby (https://stackoverflow.com/questions/51021/what-is-the-difference-between-raising-exceptions-vs-throwing-exceptions-in-ruby)) have a separate `throw` from `raise` that handle different kins of errors. (See also: https://web.archive.org/web/20210513161454/http://danielchangnyc.github.io/blog/2013/10/23/throw-raise/ ) And in other programming languages, such as [C++](https://stackoverflow.com/questions/8480640/how-to-throw-a-c-exception) and [Java](https://stackoverflow.com/questions/31215951/catching-versus-throwing-exceptions-in-java), exceptions are thrown. C++ [doesn't have a `raise`](https://www.w3schools.com/cpp/cpp_exceptions.asp)

Eg. in other ruby:

 ```
- raise and rescue are used for handling errors only
- throw and catch are used in order to terminate execution early when no other work is needed
```

[Source](https://web.archive.org/web/20201112034445/http://rubylearning.com/blog/2011/07/12/throw-catch-raise-rescue-im-so-confused/) To make matters more confusing, ruby also has a [`fail`](https://stackoverflow.com/questions/31937632/fail-vs-raise-in-ruby-should-we-really-believe-the-style-guide).

We can do that with `exit(integer)`, but we don't have much else in python. 

Then again, even the docs [get confused, slip up and say we "catch" exceptions](https://docs.python.org/3/tutorial/errors.html#:~:text=accidentally%20catching%20an%20exception%20that%20wasn%E2%80%99t%20raised%20by%20the%20code%20being%20protected%20by%20the%20try), so _why_ do we have this "special" `raise` in python, when Java, C++, [Haskell](https://stackoverflow.com/questions/6070316/how-to-throw-an-exception-and-exit-the-program-in-haskell), ruby, (and more) all have an implementation of `throw`?

Well C++ exceptions famously .... sucked at the time of writing. Consider:

``` c++
auto variable_instance = new variable() 

if (condition)
    throw ....

do_something(variable_instance)
delete variable_instance
```

which leaks `variable_instance` if `condition`. See [this SO question for more examples](https://stackoverflow.com/questions/15372242/example-of-memory-leak-in-c-by-use-of-exceptions)

Perhaps Guido wanted to distance python from this? Time will tell, [I asked a question on SO for help figuring this out](https://stackoverflow.com/staging-ground/78577134).

For now, I think we should eliminate this "Possibly Throws" title.

Reply:
I'll just change this to raises. Thanks for the info!

Reply:
Implemented this.